Signal/src/views/CaptionView.swift

```diff
@@ -0,0 +1,182 @@
+        currentCaptionView.autoPinEdgesToSuperviewEdges(with: .zero, excludingEdge: .top)
+        currentCaptionView.autoPinEdge(toSuperviewEdge: .top, withInset: 0, relation: .greaterThanOrEqual)
+
+        pendingCaptionView.alpha = 0
+        addSubview(pendingCaptionView)
+        pendingCaptionView.autoPinEdgesToSuperviewEdges(with: .zero, excludingEdge: .top)
+        pendingCaptionView.autoPinEdge(toSuperviewEdge: .top, withInset: 0, relation: .greaterThanOrEqual)
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    // MARK: Subviews
+
+    private var pendingCaptionView: CaptionView = CaptionView()
+    private var currentCaptionView: CaptionView = CaptionView()
+}
+
+private class CaptionView: UIView {
+
+    var text: String? {
+        get { return textView.text }
+
+        set {
+            if let captionText = newValue, captionText.count > 0 {
+                textView.text = captionText
+            } else {
+                textView.text = nil
+            }
+        }
+    }
+
+    // MARK: Subviews
+
+    let textView: CaptionTextView = {
+        let textView = CaptionTextView()
+
+        textView.font = UIFont.ows_dynamicTypeBody
+        textView.textColor = .white
+        textView.backgroundColor = .clear
+        textView.isEditable = false
+        textView.isSelectable = false
+
+        return textView
+    }()
+
+    let scrollFadeView = GradientView(from: .clear, to: .black)
+
+    // MARK: Initializers
+
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+
+        addSubview(textView)
+        textView.autoPinEdgesToSuperviewMargins()
+
+        addSubview(scrollFadeView)
+        scrollFadeView.autoPinEdgesToSuperviewEdges(with: .zero, excludingEdge: .top)
+        scrollFadeView.autoSetDimension(.height, toSize: 20)
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    // MARK: UIView overrides
+
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        scrollFadeView.isHidden = !textView.doesContentNeedScroll
+    }
+
+    // MARK: -
+
+    class CaptionTextView: UITextView {
+
+        var kMaxHeight: CGFloat = ScaleFromIPhone5(200)
+
+        override var text: String! {
+            didSet {
+                invalidateIntrinsicContentSize()
+            }
+        }
+
+        override var font: UIFont? {
+            didSet {
+                invalidateIntrinsicContentSize()
+            }
+        }
+
+        var doesContentNeedScroll: Bool {
+            return self.bounds.height == kMaxHeight
+        }
+
+        override func layoutSubviews() {
+            super.layoutSubviews()
+
+            // Enable/disable scrolling depending on wether we've clipped
+            // content in `intrinsicContentSize`
+            if doesContentNeedScroll {
+                if !isScrollEnabled {
```

Comment:
Why do we need this condition?

Reply:
Hmm... I don't recall. I'm not sure if setting isScrollEnabled has any non-obvious side effects I'll try removing it and see if anything seems broken - it might just be cruft.